### PR TITLE
AVRO-2693: Upgrade to netcoreapp3.1

### PR DESCRIPTION
**Pull request by zcsizmadia**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: lang/csharp/src/apache/test/CodGen/CodeGenTest.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ namespace Avro.Test
 
     class CodeGenTest
     {
-#if !NETCOREAPP2_2 // System.CodeDom compilation not supported in .NET Core: https://github.com/dotnet/corefx/issues/12180
+#if !NETCOREAPP3_1 // System.CodeDom compilation not supported in .NET Core: https://github.com/dotnet/corefx/issues/12180
```
Comment (eerhardt) [inline lang/csharp/src/apache/test/CodGen/CodeGenTest.cs]: These can be changed to just `#if !NETCOREAPP`. That way they don't have to be updated every time the TFM changes.

File: lang/csharp/src/apache/test/Avro.test.csproj [hunk 2]
```diff
@@ -17,12 +17,13 @@
...
   <PropertyGroup>
-    <TargetFrameworks Condition="'$(OS)'!='Windows_NT'">netcoreapp2.2</TargetFrameworks>
-    <TargetFrameworks Condition="'$(OS)'=='Windows_NT'">net461;netcoreapp2.2</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)'!='Windows_NT'">netcoreapp3.1</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)'=='Windows_NT'">net461;netcoreapp3.1</TargetFrameworks>
     <RootNamespace>Avro.test</RootNamespace>
     <AssemblyName>Avro.test</AssemblyName>
     <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
     <GenerateProgramFile>false</GenerateProgramFile>
+    <OutputType>Exe</OutputType>
```
Comment (eerhardt) [inline lang/csharp/src/apache/test/Avro.test.csproj]: Why is this necessary?

File: lang/csharp/src/apache/test/CodGen/CodeGenTest.cs (hunk 1 above)
Comment (zcsizmadia) [reply lang/csharp/src/apache/test/CodGen/CodeGenTest.cs]: That is correct. I wanted to keep the change to be as similar to the original as possible, however the intent is to differentiate between netcoreapp and not netcoreapp.

File: lang/csharp/src/apache/test/Avro.test.csproj (hunk 2 above)
Comment (zcsizmadia) [reply lang/csharp/src/apache/test/Avro.test.csproj]: I will double check later today, but IIRC the build failed without it using 3.1.

File: lang/csharp/src/apache/test/Avro.test.csproj [hunk 3]
```diff
@@ -42,7 +43,7 @@
     </PackageReference>
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)'!='netcoreapp2.2'">
+  <ItemGroup Condition="'$(TargetFramework)'!='netcoreapp3.1'">
```
Comment (blachniet) [inline lang/csharp/src/apache/test/Avro.test.csproj]: Which version of **Microsoft.NET.Test.Sdk** do we need to reference for **netcoreapp3.1**? We have another condition check for **netcoreapp2.2** a few lines down, on line 50. Before, the logic was

```
if netcoreapp2.2
    use v15.3.0
else
    use v15.6.1
```

I'm surprised that it's actually getting a reference to **Microsoft.NET.Test.Sdk** when targeting **netcoreapp3.1** right now with the current conditions. I must be missing something.

Comment (zcsizmadia) [reply lang/csharp/src/apache/test/Avro.test.csproj]: It is referencing the 15.6.1 version which is not the latest. The latest Test.Sdk version of 16.4.0 can be used without any netcoreapp version checking

File: lang/csharp/src/apache/test/Avro.test.csproj (hunk 2 above)
Comment (zcsizmadia) [reply lang/csharp/src/apache/test/Avro.test.csproj]: It is not needed. I messed up my original for branch, Should I redo the whole PR?

**Comment by zcsizmadia** [comment]
Recretaed PR https://github.com/apache/avro/pull/798
